src/main/java/io/supertokens/inmemorydb/queries/TOTPQueries.java

```diff
@@ -0,0 +1,181 @@
+
+        private TOTPDeviceRowMapper() {
+        }
+
+        private static TOTPDeviceRowMapper getInstance() {
+            return INSTANCE;
+        }
+
+        @Override
+        public TOTPDevice map(ResultSet result) throws SQLException {
+            return new TOTPDevice(
+                    result.getString("device_name"),
+                    result.getString("user_id"),
+                    result.getString("secret_key"),
+                    result.getInt("period"),
+                    result.getInt("skew"),
+                    result.getBoolean("verified"));
+        }
+    }
+
+    private static class TOTPUsedCodeRowMapper implements RowMapper<TOTPUsedCode, ResultSet> {
+        private static final TOTPUsedCodeRowMapper INSTANCE = new TOTPUsedCodeRowMapper();
+
+        private TOTPUsedCodeRowMapper() {
+        }
+
+        private static TOTPUsedCodeRowMapper getInstance() {
+            return INSTANCE;
+        }
+
+        @Override
+        public TOTPUsedCode map(ResultSet result) throws SQLException {
+            return new TOTPUsedCode(
+                    result.getString("user_id"),
+                    result.getString("code"),
+                    result.getBoolean("is_valid_code"),
+                    result.getLong("expiry_time"));
+        }
+    }
+
+    public static boolean insertUsedCode(Start start, TOTPUsedCode code) throws SQLException, StorageQueryException {
+        String QUERY = "INSERT INTO " + Config.getConfig(start).getTotpUsedCodesTable()
+                + " (user_id, code, is_valid_code, expiry_time) VALUES (?, ?, ?, ?);";
+
+        update(start, QUERY, pst -> {
+            pst.setString(1, code.userId);
+            pst.setString(2, code.code);
+            pst.setBoolean(3, code.isValidCode);
+            pst.setLong(4, code.expiryTime);
+        });
+        return true; // FIXME: Count the number of rows inserted OR Check if the code was inserted
+    }
+
+    public static TOTPUsedCode[] getUsedCodes(Start start, String userId) throws SQLException, StorageQueryException {
+        String QUERY = "SELECT * FROM " +
+                Config.getConfig(start).getTotpUsedCodesTable()
+                + " WHERE user_id = ?;";
+        return execute(start, QUERY, pst -> pst.setString(1, userId), result -> {
+            List<TOTPUsedCode> codes = new ArrayList<>();
+            while (result.next()) {
+                codes.add(TOTPUsedCodeRowMapper.getInstance().map(result));
+            }
+
+            return codes.toArray(new TOTPUsedCode[0]);
+        });
+    }
+
+    public static void removeExpiredCodes(Start start)
+            throws StorageTransactionLogicException, StorageQueryException, SQLException {
+        String QUERY = "DELETE FROM " + Config.getConfig(start).getTotpUsedCodesTable()
+                + " WHERE expiry_time < ?;";
+
+        update(start, QUERY, pst -> pst.setLong(1, System.currentTimeMillis()));
```

Comment:
we are doing to store expiry_time in milliseconds? If so, maybe change the column name to `expiry_time_ms`

Reply:
Okay.